src/backend/distributed/executor/executor.c

```diff
@@ -1658,36 +1665,110 @@ CheckConnectionTimeout(WorkerPool *workerPool)
 		requiredActiveConnectionCount = initiatedConnectionCount;
 	}
 
-	/*
-	 * The enforcement is not always erroring out. For example, if a SELECT task
-	 * has two different placements, we'd warn the user, fail the pool and continue
-	 * with the next placement.
-	 *
-	 * !workerPool->failed is only useful to prevent emitting WARNINGs over and over
-	 * for failed pools where failed pools have 0 ready tasks.
-	 */
-	if (activeConnectionCount < requiredActiveConnectionCount && !workerPool->failed &&
-		TimestampDifferenceExceeds(poolStartTime, now, NodeConnectionTimeout))
+	if (TimestampDifferenceExceeds(poolStartTime, now, NodeConnectionTimeout))
 	{
-		int logLevel = WARNING;
+		if (activeConnectionCount < requiredActiveConnectionCount)
+		{
+			int logLevel = WARNING;
 
-		/*
-		 * First fail the pool and create an opportunity to execute tasks
-		 * over other pools when tasks have more than one placement to execute.
-		 */
-		WorkerPoolFailed(workerPool);
+			/*
+			 * First fail the pool and create an opportunity to execute tasks
+			 * over other pools when tasks have more than one placement to execute.
+			 */
+			WorkerPoolFailed(workerPool);
 
-		logLevel = execution->errorOnAnyFailure || execution->failed ? ERROR : WARNING;
+			/*
+			 * The enforcement is not always erroring out. For example, if a SELECT task
+			 * has two different placements, we'd warn the user, fail the pool and continue
+			 * with the next placement.
+			 */
+			if (execution->errorOnAnyFailure || execution->failed)
+			{
+				logLevel = ERROR;
+			}
 
-		ereport(logLevel, (errcode(ERRCODE_CONNECTION_FAILURE),
-						   errmsg("could not establish any connections to the node "
-								  "%s:%d after %u ms", workerPool->node->workerName,
-								  workerPool->node->workerPort,
-								  NodeConnectionTimeout)));
+			ereport(logLevel, (errcode(ERRCODE_CONNECTION_FAILURE),
+							   errmsg("could not establish any connections to the node "
+									  "%s:%d after %u ms", workerPool->node->workerName,
+									  workerPool->node->workerPort,
+									  NodeConnectionTimeout)));
+		}
+		else
+		{
+			/* stop interrupting WaitEventSetWait for timeouts */
+			workerPool->survivedTimeout = true;
```

Comment:
Totally optional: I find slightly hard to follow the calculation and the name of `survivedTimeout `. It might be easier to read if we rename the flag to `checkForPoolTimeout` (or such) and revert the value of the flag.

In that case, we could even use that flag to early return in this function. Because with this change we're actually executing some of the checks of the function more often.